### PR TITLE
632 model program version2

### PR DESCRIPTION
**Pull request by Castronova**
Version 2 of the Model Program resource

* Metadata terms renaming and consolidation
* JQuery multi-select boxes to enable multi-value metadata fields
* Added events for resource submit-success and submit-failure
* Enhanced metadata validation
* Added a datetime widget to make sure releaseDate metadata term is in the correct format

### CODE REVIEW

**Comment by hydrobuild** [comment]
Can one of the admins verify this patch?

File: hs_model_program/models.py [hunk 1]
```diff
@@ -149,20 +124,50 @@ def get_xml(self):
...
-                  'software_repo',
-                  'release_notes',
-                  'user_manual',
-                  'theoretical_manual',
-                  'source_code']
-
-        model_program_object = self.program
-        self.add_metadata_element_to_xml(container, model_program_object, fields)
+        if self.program:
+            model_engine = etree.SubElement(container, '{%s}modelEngine' % self.NAMESPACES['hsterms'])
```
Comment (pkdash) [inline hs_model_program/models.py]: May be better to write a helper function that takes the list object (engines, software, documentation, releaseNotes) to generate the xml elements.

File: hs_model_program/static/css/model_program.css [hunk 2]
```diff
@@ -0,0 +1,56 @@
+
+.mpElement{
+    font-weight: bold;
+    padding: 2px 2px 2px 2px;
```
Comment (mauriel133) [inline hs_model_program/static/css/model_program.css]: Better to change to 'padding: 2px;' in this case.

File: hs_model_program/static/css/model_program.css [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    /*content:attr(title);*/
+    font-weight:bold;
+    height:2px;
+    color:transparent;
+    overflow:hidden;
+    visibility:hidden;
+}
+
+div#mp.row {
+    padding: 5px 0px 5px 0px
```
Comment (mauriel133) [inline hs_model_program/static/css/model_program.css]: Can be rewritten as 'padding: 5px 0;'. Also, if the value is 0 you don't need to specify the unit. In this case '0px' would be '0'

File: hs_model_program/static/js/model_program.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            }
+
+            // set the initial value for the date picker
+            var date = new Date(data['date_released']);
+            if (isNaN(date) == false)
+                $('#modelReleaseDate_picker').datepicker('setDate', date);
+
+        },
+        error: function (data) {
+            console.log('There was an error with model instance GET.')
```
Comment (mauriel133) [inline hs_model_program/static/js/model_program.js]: Wouldn't it be better displaying an error message in the page rather than outputting to the browser's console?

File: hs_model_program/static/js/model_program.js [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        },
+        complete: function (data) {
+            // once the selected fields (html) have been updated, build the jQuery multiselect boxes
+
+            // destroy any existing multiselect widgets
```
Comment (mauriel133) [inline hs_model_program/static/js/model_program.js]: Change this comment to 'destroy current multiselect widget' to avoid confussion

File: hs_model_program/static/css/model_program.css [hunk 6]
```diff
@@ -0,0 +1,56 @@
+
+.mpElement{
+    font-weight: bold;
+    padding: 2px 2px 2px 2px;
+}
+
+a.mpElement::after {
+    display:block;
```
Comment (mauriel133) [inline hs_model_program/static/css/model_program.css]: When writing css, try to follow the formatting we are using. For example, use a space after colon.

File: hs_model_program/static/css/model_program.css [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+div#mp.row {
+    padding: 5px 0px 5px 0px
+}
+
+option:checked{
+    background: #0073ea;
+}
+table{
```
Comment (mauriel133) [inline hs_model_program/static/css/model_program.css]: Same as above, try to leave a space between the selector and '{'

File: hs_model_program/models.py (hunk 1 above)
Comment (Castronova) [inline hs_model_program/models.py]: @pabitra I added a function to build the xml for multi-element metadata objects.  8f19d273baae438874214ed68c90c535cb27218b

File: hs_model_program/static/css/model_program.css (hunk 7 above)
Comment (Castronova) [inline hs_model_program/static/css/model_program.css]: Fixed spacing between selectors and '{'.  817e36322f124c70d4a82767b6ff9cad46ef851b

File: hs_model_program/static/css/model_program.css (hunk 6 above)
Comment (Castronova) [inline hs_model_program/static/css/model_program.css]: Fixed spacing between colons and values.  817e36322f124c70d4a82767b6ff9cad46ef851b

File: hs_model_program/static/css/model_program.css (hunk 3 above)
Comment (Castronova) [inline hs_model_program/static/css/model_program.css]: fixed this in 817e36322f124c70d4a82767b6ff9cad46ef851b

File: hs_model_program/static/css/model_program.css (hunk 2 above)
Comment (Castronova) [inline hs_model_program/static/css/model_program.css]: fixed this in 817e36322f124c70d4a82767b6ff9cad46ef851b

File: hs_model_program/static/js/model_program.js (hunk 5 above)
Comment (Castronova) [inline hs_model_program/static/js/model_program.js]: This comment has been changed in 9f54f809f8d2d723d140ee65ca6e4f4b6aa20673

File: hs_model_program/static/js/model_program.js (hunk 4 above)
Comment (Castronova) [inline hs_model_program/static/js/model_program.js]: I replaced the console message with an error message that appears on the resource edit page. b6c99f237b32c2513792b08e9828a379ddc05d77

File: hs_core/templates/pages/genericresource.html [hunk 8]
```diff
@@ -2644,7 +2644,8 @@ <h3 style="color: red" id="file-type-error"></h3>
                         json_response = JSON.parse(result);
                         if (json_response.status === 'success')
                         {
-                            $form.find("button").hide();
+                            $(document).trigger(jQuery.Event("submit-success"));
```
Comment (ironfroggy) [inline hs_core/templates/pages/genericresource.html]: You can pass the name of the event directly to `trigger()` and skip `jQuery.Event()`

File: hs_model_program/forms.py [hunk 9]
```diff
@@ -12,18 +12,47 @@
...
     def __init__(self, allow_edit=True, res_short_id=None, element_id=None, element_name=None, *args, **kwargs):
 
+        files = kwargs.pop('files')
+        file_data = {}
+        for f in files.all():
+            short_url = f.resource_file.name
+            name = short_url.split('/')[-1]
+            file_data[name] = short_url
+
+        options = '\n'.join(['<option value=%s>%s</option>'%(value, key) for key, value in file_data.iteritems() ])
```
Comment (ironfroggy) [inline hs_model_program/forms.py]: There is no escaping of these strings, which why generating HTML outside of templates is something you need to be so careful about.

File: hs_model_program/static/js/model_program.js [hunk 10]
```diff
@@ -0,0 +1,179 @@
+/**
+ * Created by tonycastronova on 10/12/15.
+ */
+
+$alert_error = '<div class="alert alert-danger" id="error-alert"> \
+                    <button type="button" class="close" data-dismiss="alert">x</button> \
+                    <strong>Error </strong> \
+                    failed to retrieve metadata for this resource.\
+                </div>'
```
Comment (ironfroggy) [inline hs_model_program/static/js/model_program.js]: The function `metadata_update_ajax_submit()` changes this variable, and thus the error that will be used after it is called. Is that he behavior you really want?

File: hs_model_program/static/js/model_program.js [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+ * Sets the hidden date field base on the datepicker.  The hidden field is what is submitted.
+ */
+function onChange_releaseDate(event){
+    var date = document.getElementById('id_modelReleaseDate');
+    var date_picker = document.getElementById("modelReleaseDate_picker");
+    date.value = date_picker.value;
+
+    // activate the save button
+    $("#resourceSpecificTab").find('.btn-primary').show();
+}
```
Comment (ironfroggy) [inline hs_model_program/static/js/model_program.js]: if this is only used above then it should be part of the call to `datepicker()` inline

File: hs_model_program/static/js/model_program.js [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+    // call django view to get model program metadata files
+    $.ajax({
+        type: "GET",
+        url: '/hsapi/_internal/get-model-metadata/',
+        data: {resource_id: shortid},
+        success: function (data) {
+
+            // get the fieldset items
+            var fs = document.getElementById("id-mpmetadata").getElementsByTagName('fieldset')[0];
```
Comment (ironfroggy) [inline hs_model_program/static/js/model_program.js]: would be easier to read a single query here, either with `$()` or `document.querySelector()`

File: hs_model_program/static/js/model_program.js [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    loadWidgets();
+});
+
+/**
+ * reloads the jQuery multi-select widgets on form submission failure
+ */
+$(document).bind("submit-error", function(event){
+    // reload dynamically generated widgets
+    loadWidgets();
+});
```
Comment (ironfroggy) [inline hs_model_program/static/js/model_program.js]: You can pass a list of events if they have the same handler. Also, `bind()` is replaced with `on()`.

`$(document).on("submit-error submit-success", ...)`

File: hs_model_program/static/js/model_program.js [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+    var parent = $(e.currentTarget.parentElement);
+
+    // get selected items from select list
+    var selected = parent.find('#mp-multi-select').find("option:selected");
+
+    // objects to store names and values in hidden metadata fields
+    var values = [];
+
+    // grab all of the selected values
+    for(var i=0; i<selected.length; i++){
```
Comment (ironfroggy) [inline hs_model_program/static/js/model_program.js]: white space after the `for` please

File: hs_model_program/static/js/underscore.js [hunk 15]
```diff
@@ -1,1276 +0,0 @@
...
-    },
-
-    // Extracts the result from a wrapped and chained object.
-    value: function() {
-      return this._wrapped;
-    }
-
-  });
-
-}).call(this);
```
Comment (ironfroggy) [inline hs_model_program/static/js/underscore.js]: It definitely looks like this library is being used. Why is it being removed?

File: hs_model_program/urls.py [hunk 16]
```diff
@@ -3,8 +3,9 @@
...
 from hs_model_program import views
+from django.contrib.staticfiles.urls import staticfiles_urlpatterns
 
 urlpatterns = patterns('',
     url(r'^_internal/get-model-metadata/$', views.get_model_metadata),
-
-    )
\ No newline at end of file
+    )
+urlpatterns += staticfiles_urlpatterns()
```
Comment (ironfroggy) [inline hs_model_program/urls.py]: this should only exist in the project top level `urls.py`

File: hs_core/templates/pages/genericresource.html (hunk 8 above)
Comment (Castronova) [inline hs_core/templates/pages/genericresource.html]: addressed in 8728905d01b2ec7936d9ade61faf7b055fbe76e2

File: hs_model_program/static/js/model_program.js (hunk 10 above)
Comment (Castronova) [inline hs_model_program/static/js/model_program.js]: I have changed the variable to `mp_alert_error` so that its value will not be modified outside this js code.  efd0967e01f4135086f657fc430554bd69b59f1c

File: hs_model_program/static/js/model_program.js (hunk 11 above)
Comment (Castronova) [inline hs_model_program/static/js/model_program.js]: Moved this function into the inline.  451a1d08beca2b193a7cc1ce264d20fefd48f2af

File: hs_model_program/static/js/model_program.js (hunk 12 above)
Comment (Castronova) [inline hs_model_program/static/js/model_program.js]: This has been replaced with a jQuery selector 8a1e045f5e011aedd64c7572422b62de34b7e402

File: hs_model_program/static/js/model_program.js (hunk 13 above)
Comment (Castronova) [inline hs_model_program/static/js/model_program.js]: These functions have been consolidated 9fac0f54c6aac35d816e52ec231c8020bd08807b

File: hs_model_program/static/js/underscore.js (hunk 15 above)
Comment (Castronova) [inline hs_model_program/static/js/underscore.js]: I've restored this file  898ada7f2ac3c6c785e6a5ad7984373c5a2b3017

File: hs_model_program/static/js/model_program.js (hunk 14 above)
Comment (Castronova) [inline hs_model_program/static/js/model_program.js]: fixed 898ada7f2ac3c6c785e6a5ad7984373c5a2b3017

File: hs_model_program/urls.py (hunk 16 above)
Comment (Castronova) [inline hs_model_program/urls.py]: This has been removed  e7e75ec751c3bcf201fb1adbf55267de46750532

File: hs_model_program/forms.py (hunk 9 above)
Comment (Castronova) [inline hs_model_program/forms.py]: @ironfroggy Can you explain this to me a little?  Would the django escape function help out here?

e.g.
```
from django.utils.html import escape
escape(some_string)
```

**Comment by hyi** [comment]
@Castronova Can you summarize what is left to be done for this PR and who needs to do the final review to give a +1 before it can be merged?

**Comment by Castronova** [comment]
@hyi The only issue remaining is to handle string escaping when generating html in `forms.py`.  I'm not sure how to address this and I am waiting to hear back from @ironfroggy.  Do you have any thoughts?

**Comment by hyi** [comment]
@Castronova What you proposed to use django.utils.html escape() to create html text looks correct to me, but I'll defer to @ironfroggy for the final words. @ironfroggy could you give a quick scan-through to see whether you can give a +1 for this PR?

File: hs_model_program/forms.py (hunk 9 above)
Comment (Castronova) [inline hs_model_program/forms.py]: Added url string escaping in commit 75fc26f7dee3bb54dbdb709a2dc495378155fbeb

**Comment by ironfroggy** [comment]
The `escape()` solution is fine. +1
